### ISSUE TITLE
Code Review (?)

### ISSUE THREAD

**Issue by md5sha256** [lero4ka16/te4j#1]
https://github.com/lero4ka16/te4j/blob/512eed9cecbb8c651b34282545d74fdc36c74beb/src/main/java/com/github/lero4ka16/te4j/util/RuntimeJavaCompiler.java#L68

You should probably just make the method synchronized instead of having a sync block on `TMP` as I assume `#compile` should only be accessed by one thread at any given moment anyway. 

https://github.com/lero4ka16/te4j/blob/512eed9cecbb8c651b34282545d74fdc36c74beb/src/main/java/com/github/lero4ka16/te4j/util/RuntimeJavaCompiler.java#L50-L66
This isn't thread safe! You synchronize the `TMP` variable but your `superclass`, `interfaces` and `content` variables can be concurrently modified during compilation!

https://github.com/lero4ka16/te4j/blob/512eed9cecbb8c651b34282545d74fdc36c74beb/src/main/java/com/github/lero4ka16/te4j/util/RuntimeJavaCompiler.java#L100-L126

What are you trying to do here? If you're validating the generated classes just check if the compiler failed. So: 
Based on Java 8 [documentation](https://docs.oracle.com/javase/8/docs/api/javax/tools/Tool.html) for `JavaCompiler`

```java
int code = compiler#run(...)
if (code != 0) {
   // some error in compilation
}
```

**Comment by lero4ka16** [comment]
Each template compilation creates `TemplateCompileContext` which creates `RuntimeJavaCompiler`.
The synchronization for `superclass`, `interfaces` and `content` is unnecessary because these fields modified by only one thread (the compilation thread)

The synchronization with TMP is necessary because compilation may start in different threads and output will written to the file `tmp/GeneratedTemplate.java` and if I run many threads which compiles a template, the file will be corrupted.

(btw sorry if my english is bad)

**Comment by md5sha256** [comment]
> The synchronization with TMP is necessary because compilation may start in different threads and output will written to the file tmp/GeneratedTemplate.java and if I run many threads which compiles a template, the file will be corrupted.

Fair enough with regards to the synchronization on `TMP`, I missed that it was a `static` field. 

> The synchronization for superclass, interfaces and content is unnecessary because these fields modified by only one thread (the compilation thread)

I understand you did not intend for the compiler to be used directly as it is probably internal; however, the constructors and methods are all public. In essence, there is a potential for misuse. 

Additionally, looking at [f8d4d8b](https://github.com/lero4ka16/te4j/commit/f8d4d8b70b93b1937490b73f4fee76a2799dd1d7), why not just use the locks which java has bulitin?

https://github.com/lero4ka16/te4j/blob/f8d4d8b70b93b1937490b73f4fee76a2799dd1d7/src/main/java/com/github/lero4ka16/te4j/util/RuntimeJavaCompiler.java#L110
One last thing, you could probably just cache the `URL[]` you supply when generating a new classloader as it seems `new URL[]{TMP.toURI().toURL()}` is constant. Not sure how much performance you'd gain but at least you don't have to instantiate the objects every time you compile .

**Comment by lero4ka16** [comment]
> One last thing, you could probably just cache the URL[] you supply when generating a new classloader as it seems new URL[]{TMP.toURI().toURL()} is constant. Not sure how much performance you'd gain but at least you don't have to instantiate the objects every time you compile .

toURL() throws an MalformedURLException and I cannot just add a static field

>  you'll be compiling different templates in parallel.
It is unnecessary I think because templates won't change too often

> why not just use the locks which java has bulitin?
Are you about ReentrantLock?

The Lock class used in `Template.HotReloadingWrapper` and when I run `Template#renderAsBytes`, `Template#renderAsString` or `Template#renderTo` it checks for locked. If locked it runs `wait`

If I replace it to ReentrantLock I can't run these methods concurrently

**Comment by md5sha256** [comment]
> toURL() throws an MalformedURLException and I cannot just add a static field
That's why static blocks exist. That being said, not sure if you want to fail-fast if the URLs are malformed.

**Comment by lero4ka16** [comment]
It's all?

**Comment by md5sha256** [comment]
Yep. Good luck!
